packages/flutter/lib/src/gestures/tap.dart

```diff
@@ -183,13 +184,30 @@ abstract class BaseTapGestureRecognizer extends PrimaryPointerGestureRecognizer
 
   @override
   void addAllowedPointer(PointerDownEvent event) {
+    assert(event != null);
     if (state == GestureRecognizerState.ready) {
       // `_down` must be assigned in this method instead of `handlePrimaryPointer`,
       // because `acceptGesture` might be called before `handlePrimaryPointer`,
       // which relies on `_down` to call `handleTapDown`.
       _down = event;
     }
-    super.addAllowedPointer(event);
+    if (_down != null) {
+      // This happens when this tap gesture has been rejected while the pointer
+      // is down (i.e. due to movement), when another allowed pointer is added,
+      // in which case all pointers are simply ignored. The `_down` being null
+      // means that _reset() has been called, since it is always set at the
+      // first allowed down event and will not be cleared except for reset(),
+      super.addAllowedPointer(event);
+    }
+  }
+
+  @override
+  @protected
+  void startTrackingPointer(int pointer, [Matrix4 transform]) {
+    // If the recognizer is tracked, it can be accepted and called [onTap],
```

Comment:
I don't understand this comment. recognizers are not tracked, pointers are. And what do you mean by "called [onTap]"?

Reply:
I changed this comment to 
```
    // The recognizer should never track any pointers when `_down` is null,
    // because calling `_checkDown` in this state will throw exception.
```

Let me know if this still sounds unclear.